File: parsec/api/protocole/events.py
# Parsec Cloud (https://parsec.cloud) Copyright (c) AGPLv3 2019 Scille SAS

from parsec.serde import OneOfSchema, fields, validate
from parsec.api.protocole.base import BaseReqSchema, BaseRepSchema, CmdSerializer


EVENTS = ("pinged", "realm.vlobs_updated", "message.received")


class EventsPingedRepSchema(BaseRepSchema):
    status = fields.CheckedConstant("ok", required=True)
    event = fields.CheckedConstant("pinged", required=True)
    ping = fields.String(validate=validate.Length(max=64), required=True)


class EventsRealmVlobsUpdatedRepSchema(BaseRepSchema):
    status = fields.CheckedConstant("ok", required=True)
    event = fields.CheckedConstant("realm.vlobs_updated", required=True)
    realm_id = fields.UUID(required=True)
    checkpoint = fields.Integer(required=True)
    src_id = fields.UUID(required=True)
    src_version = fields.Integer(required=True)


class EventsRealmMaintenanceStartedRepSchema(BaseRepSchema):
    status = fields.CheckedConstant("ok", required=True)
    event = fields.CheckedConstant("realm.maintenance_started", required=True)
    realm_id = fields.UUID(required=True)
    encryption_revision = fields.Integer(required=True)


class EventsRealmMaintenanceFinishedRepSchema(BaseRepSchema):
    status = fields.CheckedConstant("ok", required=True)
    event = fields.CheckedConstant("realm.maintenance_finished", required=True)
    realm_id = fields.UUID(required=True)
    encryption_revision = fields.Integer(required=True)


class EventsMessageReceivedRepSchema(BaseRepSchema):
    status = fields.CheckedConstant("ok", required=True)
    event = fields.CheckedConstant("message.received", required=True)
    index = fields.Integer(required=True)


class EventsListenReqSchema(BaseReqSchema):
    wait = fields.Boolean(missing=True)


class EventsListenRepSchema(OneOfSchema):
    type_field = "event"
    type_field_remove = False
    type_schemas = {
        "pinged": EventsPingedRepSchema(),
        "realm.vlobs_updated": EventsRealmVlobsUpdatedRepSchema(),
        "realm.maintenance_started": EventsRealmMaintenanceStartedRepSchema(),
        "realm.maintenance_finished": EventsRealmMaintenanceFinishedRepSchema(),
        "message.received": EventsMessageReceivedRepSchema(),
    }

    def get_obj_type(self, obj):
        return obj["event"]


events_listen_serializer = CmdSerializer(EventsListenReqSchema, EventsListenRepSchema)


class EventsSubscribeReqSchema(BaseReqSchema):
    ping = fields.List(fields.String(validate=validate.Length(max=64)), missing=None)
    realm = fields.List(fields.UUID(), missing=None)
    message = fields.Boolean(missing=None)


class EventsSubscribeRepSchema(BaseRepSchema):
    pass


events_subscribe_serializer = CmdSerializer(EventsSubscribeReqSchema, EventsSubscribeRepSchema)

File: oxidation/crates/parsec_api_protocol/src/events.rs
// Parsec Cloud (https://parsec.cloud) Copyright (c) BSLv1.1 (eventually AGPLv3) 2016-2021 Scille SAS

use serde::{Deserialize, Serialize};

use parsec_api_types::{InvitationToken, RealmID, RealmRole, VlobID};
use parsec_schema::parsec_schema;

use crate::InvitationStatus;

/*
 * APIEvent
 */

#[parsec_schema]
#[serde(tag = "event")]
pub enum APIEvent {
    #[serde(rename = "pinged")]
    Pinged { ping: String },
    #[serde(rename = "message.received")]
    MessageReceived { index: u64 },
    #[serde(rename = "invite.status_changed")]
    InviteStatusChanged {
        token: InvitationToken,
        invitation_status: InvitationStatus,
    },
    #[serde(rename = "realm.maintenance_finished")]
    RealmMaintenanceFinished {
        realm_id: RealmID,
        encryption_revision: u64,
    },
    #[serde(rename = "realm.maintenance_started")]
    RealmMaintenanceStarted {
        realm_id: RealmID,
        encryption_revision: u64,
    },
    #[serde(rename = "realm.vlobs_updated")]
    RealmVlobsUpdated {
        realm_id: RealmID,
        checkpoint: u64,
        src_id: VlobID,
        src_version: u64,
    },
    #[serde(rename = "realm.roles_updated")]
    RealmRolesUpdated { realm_id: RealmID, role: RealmRole },
}

/*
 * EventsListenReq
 */

#[parsec_schema]
pub struct EventsListenReq {
    pub wait: bool,
}

#[parsec_schema]
#[serde(tag = "status", rename_all = "snake_case")]
pub enum EventsListenRep {
    Ok(APIEvent),
    Cancelled { reason: Option<String> },
    NoEvents,
    UnknownError { error: String },
}

/*
 * EventsSubscribeReq
 */

#[parsec_schema]
pub struct EventsSubscribeReq;

/*
 * EventsSubscribeRep
 */

#[parsec_schema]
#[serde(tag = "status", rename_all = "snake_case")]
pub enum EventsSubscribeRep {
    Ok,
    UnknownError { error: String },
}
